File: p13/Entry4A.java
@Author: Yash

class Ship extends GO{
  void Collide(GameObject go2){ //1
    go2.collide(this);
  }
  void Collide(Station go2){ //2
    // logic to  collide ship and station
  }
  void Collide(Commet go2){ //3
    // logic to  collide ship and Commet
  }
  void Collide(Aestroid go2){ //4
    // logic to  collide ship and Aestroid
  }
  void Collide(Ship go2){ //5
    // logic to  collide ship and Ship
  }
}

class Comet extends GO{
  void Collide(GameObject go2){
    go2.collide(this);
  }
  void Collide(Ship go2){
    // logic to  collide station and ship
  }
  void Collide(Station go2){
    // logic to  collide station and station
  }
  void Collide(Commet go2){
    // logic to  collide station and Commet
  }
  void Collide(Aestroid go2){
    // logic to  collide station and Aestroid
  }
  
}

class Handler
{
     public void Collide(GameObject go1,GameObject go2)
     {
          go1.Collide(go2);
     }
}

File: p13/Entry4A.py
class Ship:
    def collide(self, other):
        other.collide_with_ship(self)

    def collide_with_station(self, station):
        # logic to collide ship and station
        pass

    def collide_with_comet(self, comet):
        # logic to collide ship and comet
        pass

    def collide_with_asteroid(self, asteroid):
        # logic to collide ship and asteroid
        pass

    def collide_with_ship(self, ship):
        # logic to collide ship and ship
        pass


class Comet:
    def collide(self, other):
        other.collide_with_comet(self)

    def collide_with_ship(self, ship):
        # logic to collide comet and ship
        pass

    def collide_with_station(self, station):
        # logic to collide comet and station
        pass

    def collide_with_comet(self, comet):
        # logic to collide comet and comet
        pass

    def collide_with_asteroid(self, asteroid):
        # logic to collide comet and asteroid
        pass


class Handler:
    def collide(self, go1, go2):
        go1.collide(go2)
